autoPyTorch/api/base_task.py

```diff
@@ -1219,10 +1314,29 @@ def refit(
 
         return self
 
-    def fit(self,
-            dataset: BaseDataset,
-            pipeline_config: Optional[Configuration] = None,
-            split_id: int = 0) -> BasePipeline:
+    def fit_pipeline(
+        self,
+        configuration: Configuration,
+        dataset: Optional[BaseDataset] = None,
+        X_train: Optional[Union[List, pd.DataFrame, np.ndarray]] = None,
+        y_train: Optional[Union[List, pd.DataFrame, np.ndarray]] = None,
+        X_test: Optional[Union[List, pd.DataFrame, np.ndarray]] = None,
+        y_test: Optional[Union[List, pd.DataFrame, np.ndarray]] = None,
+        dataset_name: Optional[str] = None,
+        resampling_strategy: Optional[Union[HoldoutValTypes, CrossValTypes]] = None,
+        resampling_strategy_args: Optional[Dict[str, Any]] = None,
+        run_time_limit_secs: int = 60,
+        memory_limit: Optional[int] = None,
+        eval_metric: Optional[str] = None,
+        all_supported_metrics: bool = False,
+        budget_type: Optional[str] = None,
+        include_components: Optional[Dict] = None,
+        exclude_components: Optional[Dict] = None,
+        search_space_updates: Optional[HyperparameterSearchSpaceUpdates] = None,
+        budget: Optional[float] = None,
+        pipeline_options: Optional[Dict] = None,
+        disable_file_output: Optional[List[str]] = None,
```

Comment:
Why do we need that many variables now?
Most information is in the BaseDataset, isn't it?
Let's try reducing variable numbers as much as possible,
because it will be painful when you create tests and it will increase line numbers and doc-string in many places.

[Clean code](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882) also says methods with 5 variables should be considered to be re-designed.
Basically, we use an object to wrap those variables in such cases.

Reply:
I agree we can get rid of some of these variables for example, `include_components`, `exclude_components` and `search_space_updates`.  I have intentionally kept the other parameters because these are needed features to be able to fit a pipeline. These parameters allow for fitting a pipeline independent of the search parameters. For example, we used a different resampling strategy (`NoResampling`) in our experiments with `reg_cocktails` which allows for fitting a pipeline with the whole train data rather than making splits with them.  We used it to get the full test accuracy of the incumbent pipeline.

Btw, dataset and X_train, y_train, X_test, y_test, dataset_name, resampling_strategy, resampling_strategy_args are exclusive to each other. The user can either send a `BaseDataset` or the data separately and we make the dataset. Would you prefer just having `dataset` so the user always has to get a dataset before fitting the pipeline?

Reply:
Btw, let me know which other parameters are not necessary for this function and we can discuss them

Reply:
> Basically, we use an object to wrap those variables in such cases.

Okay, I think we can do this. It will be helpful also for other functions like `search` and `fit_ensemble`. Do you mind if I do it in another PR? I think it is more appropriate to make a new PR so I can unify this for other functions as well.